src/meta_ops.rs

```diff
@@ -207,54 +304,50 @@ pub fn new_index<'gc>(
                     Ok(CallbackReturn::Call {
                         function: call.function,
                         then: None,
-                    }
-                    .into())
+                    })
                 } else {
                     Ok(CallbackReturn::Return)
                 }
             })
             .into(),
-            args: [table.into(), key, value],
+            args: [table, key, value],
         },
         _ => MetaCall {
-            function: call(ctx, idx)?,
+            function: call(ctx, idx)
+                .map_err(|e| MetaOperatorError::Call(MetaMethod::NewIndex, e))?,
             args: [table, key, value],
         },
     }))
 }
 
-pub fn call<'gc>(ctx: Context<'gc>, v: Value<'gc>) -> Result<Function<'gc>, TypeError> {
+pub fn call<'gc>(ctx: Context<'gc>, v: Value<'gc>) -> Result<Function<'gc>, MetaCallError> {
     let metatable = match v {
         Value::Function(f) => return Ok(f),
         Value::Table(t) => t.metatable(),
         Value::UserData(ud) => ud.metatable(),
         _ => None,
     }
-    .ok_or(TypeError {
-        expected: "function",
-        found: v.type_name(),
-    })?;
+    .ok_or(MetaCallError(v.type_name()))?;
 
     match metatable.get(ctx, MetaMethod::Call) {
         f @ (Value::Function(_) | Value::Table(_) | Value::UserData(_)) => Ok(
+            // TODO: detect call chain length, error if too long
+            // Example: t = {}; setmetatable(t, { __call = t }); t()
+            // (though PUC-Rio Lua only detects long index chains, not long call chains)
```

Comment:
Same logic as above, this is not necessary. Actually, this serves as more evidence that the MAXTAGLOOP exists solely because `__index` loops may not trigger hook functions.